### DISCUSSION TITLE
9.5.0(minimal,gz) flashed, trying to upgrade, to no avail

### DISCUSSION THREAD

**Discussion by corbolais** [arendst/Tasmota#12579] [Q&A] [answered]
Issue: https://github.com/arendst/Tasmota/issues/12568, https://github.com/arendst/Tasmota/issues/12575

Upgraded to 9.5.0(minimal), trying to upgrade to tasmota-sensors or tasmota-minimal not possible.
Looking at google search results for the error message "not enough space" did not lead to any successful upgrade.

```
status 4
{"StatusMEM":{"ProgramSize":389,"Free":1656,"Heap":31,"ProgramFlashSize":4096,"FlashSize":1024, [...]}
```

Of course, I already did try to use a tasmota-minimal.gz, it is not working. I already had other esp8266s with the same issue, the only way to fix it was reflashing via uart. This ist however not a solution that is always feasible. 

Similar: https://github.com/arendst/Tasmota/discussions/12286

edit: I mentioned in https://github.com/arendst/Tasmota/discussions/12286: 
I flashed 9.1 minimal on 8.5.1 and then 9.5 minimal on top of that (8.5 -> 9.1 minimal -> 9.5 minimal). Starting with 9.5 I cannot even reflash 9.5 minimal, be it compressed or not, self-compiled or stock.

**Comment by corbolais** [comment]
Uploads via OTA and webpage upgrade both fail.

**Reply by barbudor** [reply]
You don't give any details
- From which version and variant did you started ?
- Through which version did you went through ?

All file upload or http-ota should use **bin.gz**
All upgrade from one version to another should follow the upgrade path using tasmota-lite.bin.gz 
If you use tasmota-minimal, you MUST upgrade to full tasmota or tasmota-lite of the SAME version before going to the next version.

**Reply by corbolais** [reply]
updated description, mentioned version info in another discussion, missed to import that info.

Is there a way out of this situation? (w/o flashing via uart)

**Reply by corbolais** [reply]
Upgrading using tasmota-minimal.gz and tasmota-lite.bin.gz with every version upgrade is the way to go? I must have missed that. IMO, the Upgrade-page is overloaded. That's probably why I missed this procedure. The one thing I got is the version string that must be followed. If there's info on having to init any version upgrade with a tasmota-lite after a tasmota-minimal, it's not concise enough.

**Reply by barbudor** [reply]
![image](https://user-images.githubusercontent.com/2996042/124585858-078a7400-de56-11eb-9b16-22a44145fbc2.png)

Isn't that clear enough ?
- DO NOT
- Links for each version are pointing to the lite version
Your're welcome to suggest documentation enhancement through the pencil icon on the top right

Going from 8.5.1 to 9.x using minimal may have broken your settings.
What are your exact situation now ?
Do you have access to the web GUI ? Which version and variant ?

If you are on a minimal I would suggest trying to upload the `tasmota-lite.bin.gz` of the same version
The best - if your device has access to the Internet - would be to proceed through OTA (not file upload) using the URL `http://ota.tasmota.com/tasmota/release-X.Y.Z/tasmota-lite.bin.gz`  (replace X.Y.Z but the right numbers - note it is **http** not https)
If you can't, try file upload.

At this stage, if succeed, the safer is to clear the settings and reconfigure form scratch.
To do that in the console enter the command `reset 5` to clear all settings but wifi  (or `reset 1` but you will have to reconfigure Wifi).
**Don't do `reset 5` or `reset 1` on minimal !**

**Reply by corbolais** [reply]
It is (obviously) not clear enough. ***The*** MOST CRITICAL *textual* information FIRST. Visual information is still strong enough, so it is OK after the most critical textual information.

Thank you for the invitation to contribute. I might do so.

Thank you for trying to detail a way out of this situation. It is not working.

I don't mind broken settings. If it is connecting and upgrading, I'm fine.

Using GUI, minimal 9.5.0, `http://ota.tasmota.com/tasmota/release-9.5.0/tasmota-lite.bin.gz`:

`UPGRADE = {"Upgrade":"Failed Update error: ERROR[8]: Flash config wrong real: 1048576 IDE: 4194304"}`

Locally, I previously solved this type of issue by compiling the binary for esp8266_1M. Is there such a thing downloadable for tasmota-stock?

**Reply by ascillato2** [reply]
> ... by compiling the binary for esp8266_1M. Is there such a thing downloadable for tasmota-stock?

Yes, all files in http://ota.tasmota.com/tasmota/ are for ESP8266 and 1MB flash

Tasmotizer is the recommended tool for flashing ( https://github.com/tasmota/tasmotizer/releases/tag/v.1.2 )

**Reply by corbolais** [reply]
OK, so how can the above error be explained and how can it be solved?

**Reply by ascillato2** [reply]
> how can the above error be explained?

As https://github.com/arendst/Tasmota/discussions/12579#discussioncomment-969487, your user settings may be broken due to preforming the upgrade using **_minimal_** after **_minimal_** versions. The upgrade procedure is from a **_"Full version"_** to **_"minimal new version"_** to **_"Full new version"_** (_repecting the upgrade path stated in the docs_).

**_Under the hood explanation:_** The minimal firmware must be as small as possible, so it has no features and it will not save user data in your flash. The user data memory map sometimes changes from one version to another and the full version takes care of that detecting and UPGRADING the old map to the new map.
The minimal version is just to free space to allow installing a big version again.
Example: 1024 KB of FLASH (1MB). In the last part of the flash, less than 100K are being used for Settings and SDK. So you have about 900KB for program AND OTA. If you have a firmware of 500KB, you will have 400KB for OTA. So, in order to upgrade to a new firmware of 500KB, you need to flash first a smaller **_intermediate_** version of 380KB for example. After that, you will have a firmware of 380KB and 520KB free for OTA. and then you can OTA the Full version you want of 500KB for example. This process is called OTA-Magic and is done automagically if using OTAurl upgrade button of Tasmota WebUI. Just set the full version you want and Tasmota will do the intermediate step of minimal without the need that you explicit it.

If you want more information of this, please refer to the Tasmota documentation or older issues.

> how can it be solved?

As https://github.com/arendst/Tasmota/discussions/12579#discussioncomment-969487, upload the **_tasmota-lite.bin.gz_** of the same version you have in your device.
After that do a `reset 1` in the console. If that fails, you need to erase your flash and flash by serial.

**Reply by corbolais** [reply]
> Isn't that clear enough ?

I just realized I indeed had noticed the -lite.gz instruction. My reasoning not to follow suite was the -lite variant is even larger a binary than -minimal, so why would the -lite variant succeed where the lighter -minimal variant fails.

**Reply by ascillato2** [reply]
The lite version has the routines for UPGRADING the settings memory map.

**Reply by meingraham** [reply]
@corbolais 

Minimal has one application only - interim upgrade path between a "complete" version of the firmware. It can be used only AFTER a device has already been configured once (i.e., the settings are configured prior to an upgrade and retained during the upgrade process utilizing minimal). Minimal does NOT provide a way to set up a device. Thus, if that is the initial variant flashed on the device, you will not be able to set it up. Minimal is not for ANY normal run-time use case. You must first flash a "complete" version of Tasmota. The Lite version is the smallest pre-compiled binary variant that can be used for "regular" operation.

You may wish to be a rebel and not follow the recommendations outlined in the documentation. But you will end up being a rebel without a working device 😉

**Reply by corbolais** [reply]
> As [#12579 (reply in thread)](https://github.com/arendst/Tasmota/discussions/12579#discussioncomment-969487), upload the **_tasmota-lite.bin.gz_** of the same version you have in your device.

That's what I did: https://github.com/arendst/Tasmota/discussions/12579#discussioncomment-969594

**Reply by ascillato2** [reply]
Ok, If that fails, you need to erase your flash and flash by serial. Sorry. The memory map settings of your device is scrambled.

**Reply by ascillato2** [reply]
Please, remember to erase the flash first and flash the full tasmota version. Please, flash http://ota.tasmota.com/tasmota/tasmota.bin

**Reply by corbolais** [reply]
> You may wish to be a rebel and not follow the recommendations outlined in the documentation. But you will end up being a rebel without a working device

Such a non-sensical comment was to be expected. Look at it the other way round, if you happen to be able to wrap your head around: Reading instructions and plainly exec'ing them is easy. Every dumb... can do that. Reading and interpreting, logically verifying or applying (short: thinking, using your head) to your individual situation is another thing lesser people are able to do. Usually those referencing (judging) others by "rebels".

Of course, if there is information missed or unknown, misinterpretations may happen.

**Reply by corbolais** [reply]
> Ok, If that fails, you need to erase your flash and flash by serial. Sorry. The memory map settings of your device is scrambled.

Don't like the sound of that. No manual way/hidden console commands? Anything?

**Reply by ascillato2** [reply]
No, Sorry. Now you have a firmware that won't save anything into memory, and also that memory settings is scrambled. The firmware was not designed for that. The firmware was designed for the procedure explained in the docs. Sorry.

The minimal fimware has nothing inside. It has ONLY the Wi-Fi and OTA routines. Nothing else. It is an intermediate firmware for that step between an OLD FULL version and a NEW FULL version (https://github.com/arendst/Tasmota/discussions/12579#discussioncomment-970004)

You need to start from skratch. Full erase and Serial Flashing of full firmware.

**Reply by corbolais** [reply]
That's bad.

But, OK, thank you.

**Reply by corbolais** [reply]
What does it mean, what happened when the following appears, trying to upgrade by OTA and official tasmota URL?

`UPGRADE = {"Upgrade":"Failed Update error: ERROR[8]: Flash config wrong real: 1048576 IDE: 4194304"}`

Trying to upload a local -lite.gz or official -lite.gz fails with "Upload Failed. Not enough Space."

**Reply by ascillato** [reply]
Memory scrambled Sorry. It can do whatever in that state. Sorry.

You need to start from skratch. Full erase and Serial Flashing of full firmware.

**Reply by meingraham** [reply]
> Look at it the other way round, if you happen to be able to wrap your head around: Reading instructions and plainly exec'ing them is easy. 

Look at it this way... there have been literally tens of millions of Tasmota downloads... and these dumb, follow the instructions, sheep, have successfully flashed the devices based on the information provided. Not sure how the dumb developers of this firmware nor the millions of dumb users have ever made it since they are not capable of coherent logical thought.

> My reasoning not to follow suite was the -lite variant is even larger a binary than -minimal, so why would the -lite variant succeed where the lighter -minimal variant fails.

You **_assumed_** or "interpreted" a meaning from the terms "minimal" and "lite" used in the variant names and their respective binary sizes regardless of the procedure the documentation stated one should follow. You assumed incorrectly. Thus your reasoned and logical approach failed. Motto - be a sheep and just follow the rest of us dummies and you will succeed.

Some additional [reading material](https://www.amazon.com/How-Win-Friends-Influence-People/dp/0671027034)...

**Reply by ascillato** [reply]
Next time, To avoid issues, you can let Tasmota take care of all that instead of doing it manually.

You can go to **_Upgrade Firmware_** in Tasmota UI, and in **_OTAurl_** be sure you have http://ota.tasmota.com/tasmota/tasmota.bin.gz and just click on **_Upgrade from URL_**. Tasmota will take care of the minimal. It will perform both steps (minimal first, full after). You just need to point the full version you want and Tasmota will do the rest. It is pretty fast the download, flash, restart, download, flash, restart. In a minute or 2, you have your device updated to latest.

**Reply by corbolais** [reply]
> Next time, To avoid issues, you can let Tasmota take care of all that instead of doing it manually.
> 
> You can go to **_Upgrade Firmware_** in Tasmota UI, and in **_OTAurl_** be sure you have http://ota.tasmota.com/tasmota/tasmota.bin.gz and just click on **_Upgrade from URL_**. Tasmota will take care of the minimal. It will perform both steps. You just need to point the full version you want and Tasmota will do the rest.

Oh, this is good news. That's new to me, tasmota upgrading via minimal when using OTA. Good thing.

**Reply by ascillato** [reply]
> tasmota upgrading via minimal when using OTA

That is automatic. Just point the OTAurl of the full version.

**Reply by corbolais** [reply]
> > Look at it the other way round, if you happen to be able to wrap your head around: Reading instructions and plainly exec'ing them is easy.
> 
> Look at it this way... there have been literally tens of millions of Tasmota downloads... and these dumb, follow the instructions, sheep, have successfully flashed the devices based on the information provided. Not sure how the dumb developers of this firmware nor the millions of dumb users have ever made it since they are not capable of coherent logical thought.
> 
> > My reasoning not to follow suite was the -lite variant is even larger a binary than -minimal, so why would the -lite variant succeed where the lighter -minimal variant fails.
> 
> You **_assumed_** or "interpreted" a meaning from the terms "minimal" and "lite" used in the variant names and their respective binary sizes regardless of the procedure the documentation stated one should follow. You assumed incorrectly. Thus your reasoned and logical approach failed. Motto - be a sheep and just follow the rest of us dummies and you will succeed.
> 
> Some additional [reading material](https://www.amazon.com/How-Win-Friends-Influence-People/dp/0671027034)...

You're quite invasive, no one ever told you?... Try to improve on your personality. Try google, won't provide any links here.

**Reply by meingraham** [reply]
Pot, meet kettle.

**Comment by ascillato** [comment, marked as answer]
## Summary:

For upgrading, please, do as following:

### Please, Don't flash minimal as the first firmware.

### Please, don't update from an old version using minimal to minimal.

### Please, follow the upgrade steps https://tasmota.github.io/docs/Upgrading/

### If you see the message **_"upload buffer miscompare"_** means not enough space. Please, use minimal or lite first.

**_(Not following these steps can end up with a non-working device and you will need to erase all flash and flash by serial)_**

### UPGRADE OTAurl WAY:

**Please, don't use minimal in OTAurl**. Tasmota will take care of downloading and upgrading the minimal version automatically in-between updates. Just point to the Full version you want.

+ Upgrade from OLD Full Version -> New Full Version
   + Example: From Full 8.5.1 -> Full 9.1 -> 9.5 FULL
          + Set the following OTAurl in Tasmota UI and click on OTAUrl Upgrade
                http://ota.tasmota.com/tasmota/release-9.1.0/tasmota.bin.gz
          + (WAIT 2 to 5 minutes and don't reset the device)
          + Set the following OTAurl in Tasmota UI and click on OTAUrl Upgrade
                http://ota.tasmota.com/tasmota/tasmota.bin.gz
          + (WAIT 2 to 5 minutes and don't reset the device)

### UPGRADE MANUAL WAY:

**Please, don't update from an old version using minimal to minimal.**
+ Upgrade from OLD Full Version -> New minimal version -> New Full Version
   + Example: Full 8.5.1 -> minimal 9.1 -> Full 9.1 -> 9.5 minimal -> 9.5 FULL
+ Alternatively: Full 8.51 -> minimal 9.1 -> Lite 9.1 -> Lite 9.5

_(For more information on this topic, please refer to the Tasmota Docs at https://tasmota.github.io/docs/Upgrading/ and older issues. Thanks.)_

**Reply by corbolais** [reply]
Thank you.
